linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,455 @@
+#define MAX_THREAD_COUNT    1
+#define THREAD_LIMIT        65536
+
+#define TASK_RESULT_VALUES  \
+    TASK_NOT_USED,          \
+    TASK_WAITING,           \
+    TASK_WORKING
+
+MU_DEFINE_ENUM(TASK_RESULT, TASK_RESULT_VALUES);
+MU_DEFINE_ENUM_STRINGS(TASK_RESULT, TASK_RESULT_VALUES)
+
+#define THREADPOOL_STATE_VALUES \
+    THREADPOOL_STATE_NOT_OPEN, \
+    THREADPOOL_STATE_OPEN
+
+MU_DEFINE_ENUM(THREADPOOL_STATE, THREADPOOL_STATE_VALUES)
+
+MU_DEFINE_ENUM_STRINGS(THREADPOOL_STATE, THREADPOOL_STATE_VALUES)
+
+typedef struct THREADPOOL_TASK_TAG
+{
+    volatile_atomic int32_t task_state;
+    THREADPOOL_WORK_FUNCTION task_func;
+    void* task_param;
+} THREADPOOL_TASK;
+
+typedef struct THREADPOOL_TAG
+{
+    volatile_atomic int32_t state;
+    volatile_atomic int32_t thread_count;
+    uint32_t max_thread_count;
+    volatile_atomic int32_t quitting;
+    volatile_atomic int32_t task_count;
+
+    sem_t semaphore;
+
+    // Pool Array
+    THREADPOOL_TASK* task_array;
+    volatile_atomic int64_t insert_idx;
+    volatile_atomic int64_t consume_idx;
+    volatile_atomic int32_t task_array_count;
+
+    SRW_LOCK_HANDLE srw_lock;
+
+    THREAD_HANDLE* thread_handle_list;
+    uint32_t list_index;
+} THREADPOOL;
+
+int threadpool_work_func(void* param)
+{
+    if (param == NULL)
+    {
+        LogCritical("Invalid args: param: %p", param);
+    }
+    else
+    {
+        struct timespec ts;
+        THREADPOOL* threadpool = param;
+        int64_t current_index;
+
+        do
+        {
+            do
+            {
+                if (clock_gettime(CLOCK_REALTIME, &ts) == -1)
+                {
+                    LogError("Failure getting time from clock_gettime");
+                }
+                else
+                {
+                    // Setup the timeout for the semaphore to 2 sec
+                    ts.tv_sec += 2;
+                    if (sem_timedwait(&threadpool->semaphore, &ts) == 0)
+                    {
+                        break;
+                    }
+                }
+            } while (interlocked_add(&threadpool->quitting, 0) != 1);
+
+            if (interlocked_add(&threadpool->quitting, 0) != 1)
+            {
+                int32_t existing_count = interlocked_add(&threadpool->task_array_count, 0);
+
+                // We have a new item
+                current_index = interlocked_add_64(&threadpool->consume_idx, 0) % existing_count;
+
+                // Is this index waiting for us to complete it
+                srw_lock_acquire_shared(threadpool->srw_lock);
+                TASK_RESULT curr_task_result = interlocked_compare_exchange(&threadpool->task_array[current_index].task_state, TASK_WORKING, TASK_WAITING);
+                srw_lock_release_shared(threadpool->srw_lock);
+                if (TASK_WAITING == curr_task_result)
+                {
+                    int64_t consume_pos = interlocked_compare_exchange_64(&threadpool->consume_idx, current_index+1, current_index);
```

Comment:
>current_index+1 [](http://example.com/codeflow?start=100&length=15)

How do you prevent wrap-around when incrementing?

Do we need the same logic as block_storage_2/src/bs_ordered_op_writer.c where we have a head and tail and we need to compare them?